modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java

```diff
@@ -119,14 +126,18 @@ public class IgniteSqlValidator extends SqlValidatorImpl {
         HUMAN_READABLE_ALIASES_FOR = Collections.unmodifiableSet(kinds);
     }
 
-    /** Dynamic parameter values. */
-    private final Object[] dynamicParamValues;
+    /** Dynamic parameters state. */
+    private final Map<Integer, DynamicParamState> dynamicParameters;
 
-    /** Dynamic parameters SQL AST nodes for invariant checks - see {@link #validateInferredDynamicParameters()}. */
-    private final SqlDynamicParam[] dynamicParamNodes;
+    /**
+     * The same dynamic parameter can be used in the same SQL tree multiple types after a rewrite.
+     * (E.g. COALESCE(?0, ?1) is rewritten into CASE WHEN ?0 IS NOT NULL THEN ?0 ELSE ?1 END)
+     * We store them to check that every i-th parameter has the same type.
+     */
+    private final IdentityHashMap<SqlDynamicParam, SqlDynamicParam> dynamicParamNodes = new IdentityHashMap<>();
```

Comment:
Can IntMap be used here?

Reply:
No we, do not. As using int -> sql node here defeats the purpose of this map. (see the comment)
Because Validator rewrites COALESCE into a case expressions that uses the same parameter multiple times.